pkg/edgehub/controller_test.go

```diff
@@ -0,0 +1,820 @@
+		for name, module := range modules {
+			testContext.AddModule(name)
+			testContext.AddModuleGroup(name, module.Group())
+		}
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			ehc.pubConnectInfo(tt.isConnected)
+			t.Run("TestMessageContent", func(t *testing.T) {
+				msg, err := testContext.Receive(core.TwinGroup)
+				if err != nil {
+					t.Errorf("Error in receiving message from twin group: %v", err)
+				} else if msg.Content != tt.content {
+					t.Errorf("TestPubConnectInfo() Content of message received in twin group : %v, want: %v", msg.Content, tt.content)
+				}
+			})
+		})
+	}
+}
+
+//TestPostUrlRequst() tests the request sent to the placement URL and its corresponding response
+func TestPostUrlRequst(t *testing.T) {
+	tests := []struct {
+		name          string
+		fields        Controller
+		client        *http.Client
+		want          string
+		expectedError error
+	}{
+		{"TestController_postUrlRequst: Proper input ", Controller{
+			config: &config.ControllerConfig{
+				PlacementURL: TestServer.URL + "/proper_request",
+				ProjectID:    "foo",
+				NodeID:       "bar",
+			},
+		}, &http.Client{},
+			"ws://127.0.0.1:20000/foo/bar/events", nil},
+
+		{"TestController_postUrlRequest: Bad Request", Controller{
+			config: &config.ControllerConfig{
+				PlacementURL: TestServer.URL + "/bad_request",
+				ProjectID:    "foo",
+				NodeID:       "bar",
+			},
+		}, &http.Client{}, "", fmt.Errorf("bad request")},
+
+		{"TestController_postUrlRequst: Wrong URL", Controller{
+			config: &config.ControllerConfig{
+				PlacementURL: TestServer.URL + "/wrong_url",
+				ProjectID:    "foo",
+				NodeID:       "bar",
+			},
+		}, &http.Client{}, "ws://127.0.0.1:20000/foo/bar/events", nil},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			got, err := ehc.postURLRequst(tt.client)
+			if !reflect.DeepEqual(err, tt.expectedError) {
+				t.Errorf("Controller.postUrlRequst() error = %v, expectedError %v", err, tt.expectedError)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("Controller.postUrlRequst() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+//TestGetCloudHubUrl() tests the procurement of the cloudHub URL from the placement server
+func TestGetCloudHubUrl(t *testing.T) {
+	tests := []struct {
+		name            string
+		fields          Controller
+		webSocketConfig config.WebSocketConfig
+		want            string
+		expectedError   error
+	}{
+		{"TestController_getCloudHubUrl: Proper Request", Controller{
```

Comment:
is it necessary to have every test case name prefixed with the test case ? won't go test describe the test which failed ?
we could have short names like "get valid cloud hub url". this applies to all tests throughout which have such descriptions.